### DISCUSSION TITLE
hydra-zen: validating Hydra app interfaces with beartype

### DISCUSSION THREAD

**Discussion by rsokl** [beartype/beartype#64] [Show and tell]
Hello @leycec  and the rest of the beartype family!

I am very excited to share with you something that has been percolating, then brewing, and now is piping hot and ready to sip ☕. I speak of none other than [the release of hydra-zen v0.3.0](https://mit-ll-responsible-ai.github.io/hydra-zen/changes.html#v0-3-0)!

### What?

Hydrogen? Hyper-zen? Spider Gwen? hydra-zen!

Hydra and hydra-zen help standardize the process of -- and remove the boilerplate associated with -- making your projects configurable, reproducible, and scalable. Hydra is becoming widely used by the ML community (and others!) to design/run experiments, and hydra-zen hopes to make it **much** nicer to use Hydra. We bring the Zen of Python to Hydra.

>"**WHY** is this on my beartype board? I **HATE** that I am reading about things that  aren't beartype!" 
> -a mild-mannered beartype discussion-board browser

I promise that there will be bears soon! Let me mollify you... here is a bear crushing a city.

![image](https://user-images.githubusercontent.com/29104956/139108379-ab365797-cb2f-43fc-8114-a2cfa3175edb.png)

### The Problem
One of the things that [Hydra sorely lacks in is: runtime type validation](https://mit-ll-responsible-ai.github.io/hydra-zen/explanation/type_refinement.html) of the values that you configured your oh-so-precious experiment with. Well, no longer!

### The Solution
hydra-zen now offers [unabashed support](https://mit-ll-responsible-ai.github.io/hydra-zen/generated/hydra_zen.third_party.beartype.validates_with_beartype.html#hydra_zen.third_party.beartype.validates_with_beartype) for infusing your app's configurable interfaces with beartype's runtime type checking. Configuring some PyTorch objects that only give (type)hints, but make no promises? No problem! We'll auto-wrap those with `@beartype` during the configuration process.

We even have a [How-To guide](https://mit-ll-responsible-ai.github.io/hydra-zen/how_to/beartype.html) dedicated teaching readers how to inject beartype-checking to their app's interface!

### That's All, Folks!

I've been eager to share this with you, but have been waiting until the time is right (i.e. until our docs stopped being trash and `v0.3.0` was finally released). I would love to get your thoughts, cheers, and jeers!

If anyone wants to see a *really* cool rendition of a bear, and to watch the "behind the scenes" of the creation of this feature in hydra-zen, check out this PR: https://github.com/mit-ll-responsible-ai/hydra-zen/pull/128.

**Comment by leycec** [comment]
*omg omg omg omg omg omg omg omg omg omg omg*

<sup>↑↑↑↑↑ ...yup, I manually typed that for added emphasis.</sup>

It's my favourite MIT Ryan! I'm bouncing with giddiness in a log cabin.

I'm also incoherent with sleep deprivation, because I spent the evening [resolving #62 by writing up a new *"Type Hint Connectives"* subsection of our countably infinite documentation.](https://github.com/beartype/beartype/#type-hint-connectives) You're (prolly) gonna love it, based on the based content above – especially that bear face crushing human civilization. It's cuddly, but it's deadly. It's grim, but it's limned with joy.

And they call it... **Bearthusula.** :bear: :city_sunrise: 

## This Beast Has Many Heads

*Back to you.* [Hydra-Zen](https://github.com/mit-ll-responsible-ai/hydra-zen) looks phenomenal. So does the documentation, which shames me by being comprehensible and sanely distributed across multiple pages. Also, this:

> Encyclopedia Hydrazennica.

I chuckled even as I almost passed out.

## Give and Ye Shall Receive Internet Glory

*Back to you.* I'd be eternally indebted if you'd consent to my advertising Hydra-Zen on [our downstream projects icon splash bar thingy](https://github.com/beartype/beartype/blob/main/README.rst). In Markdown, this paean to power-law dynamics currently resembles:

> Beartype [powers quality assurance across the Python ecosystem](https://libraries.io/pypi/beartype/dependents).
> ![ms gonna ms](https://user-images.githubusercontent.com/217028/138580730-dcb3ee22-6372-4a2d-93bf-d1a3c97e67cf.png) ![pycrisp is crispy](https://user-images.githubusercontent.com/217028/138580797-1f09f0b4-bced-4651-82c9-6dae50afe883.png) ![sentipy is sentimental](https://user-images.githubusercontent.com/217028/138581059-6e3218d0-2a2e-42b7-b15c-7e2759fa2399.png)

You see the problem, I trust. **Hydra-Zen isn't there.** This shall not stand!

Our bear-friendly logo adores your own hydra-infested logo, but wonders if you might have a 32x32 *or* 64x64 *or* `something`x`something` PNG icon with transparent background cropping just that horrifying (but oddly enticing) hydra? So, the leftmost one-fifth of this with the white supplanted with transparency maybe:

![hydra will eat us all](https://raw.githubusercontent.com/mit-ll-responsible-ai/hydra-zen/main/brand/Hydra-Zen_logo_full_filled_bkgrnd_small.png)

If so, I'll insert you immediately after Microsoft. Sorry, [pycrisp](https://gitlab.com/pycrisp/pycrisp). The hydra takes precedence. It looks like the sort of thing that'll eat me if I don't do what it says.

**Reply by rsokl** [reply]
Things are busy busy busy! I am giving a talk at PyData Global tomorrow -- which will feature a shoutout to beartype!! 

I'll have to hold off on a full reply, but in the meantime, all of our current branding-images live here: https://github.com/mit-ll-responsible-ai/hydra-zen/tree/main/brand

`hydra_zen_favicon_64x64.ico` might be for you (I think you can just convert to png??)

In terms of the messaging, I am not exactly sure how to phrase it but hydra-zen *facilitates* the use of beartype, which is a bit different than being a satisfied beartype user! It is 100% up to you what you want to say/how to say it -- I am just brainstorming here -- but perhaps indicating that beartype can be _injected_ into Hydra-based applications is newsworthy? Totally up to you though! 😄😄

**Comment by leycec** [comment]
> I am giving a talk at PyData Global tomorrow -- which will feature a shoutout to beartype!!

![*squee*](https://memegenerator.net/img/instances/48885291/squee.jpg)

> all of our current branding-images live here

**Ho, ho, ho.** You must be the many-headed Saint Nick I once dreaded but now welcome, because [your wondrous favicon](https://github.com/mit-ll-responsible-ai/hydra-zen/blob/main/brand/Hydra-Zen%20-%20favicon.png) is *perfect* for our needs – responsibly rescaled from 128x128 down to 64x64 to prevent Microsoft from gettin' all [jelly](https://www.urbandictionary.com/define.php?term=jelly).

Relatedly, [GitHub has this questionable but quasi-useful trick of reliably hosting arbitrary assets on GitHub's CDN by embedding those assets in closed GitHub issues](https://github.com/beartype/beartype-assets/issues/5). For robustness against 404 breakage, I'm stashing the third-party icons splashed across our branding tower via this ~~awful~~ *wonderful* scheme. This now includes you, Hydra-Zen.

*Behold!* Hydra-Zen's glorious 64x64-rescaled favicon hosted on the GitHub CDN:

![behold, it rears](https://user-images.githubusercontent.com/217028/139522423-d987d484-425b-4e70-a0e7-7390593471f1.png)

I can't be proud. But I'll still do it.

> I am not exactly sure how to phrase it but hydra-zen *facilitates* the use of beartype...

That's even more exciting. Are we talkin' [reverse-dependency-injection sorta thing](https://en.wikipedia.org/wiki/Dependency_injection) or more standard-plugin-automation sorta thing? Both are cool. You know. I only ask as I'd actually *love* to leverage Hydra-Zen in our team's inevitable next CLI. The `argparse.ArgumentParser` API makes me want to rub chalk into my eyes, whereas [Click](https://click.palletsprojects.com) rocks the POSIX Casbah but fails to integrate the CLI APIs it auto-generates with configuration files – which is *always* the pain point. Just "Ugh!"

A devop's gotta ask, so: "Would there be any meaningful way for Hydra-Zen to gradually co-opt the [Click API](https://click.palletsprojects.com/en/8.0.x/#api-reference) into itself as, say, a new `hydra_zen.click` submodule?" Here at `beartype`, we're fans of devouring the competition by assimilating their lesser APIs into our own hegemonic API that consumes their userbase from the inside out even as we copy-pasta the UX design our competitors invested multiple years and hair follicles into.

I can't be proud. But I'll still do it.

Lastly... [**Hydra-Zen is now featured in our branding tower**](https://github.com/beartype/beartype/blob/main/README.rst). Let's leave this issue open and visible for a bit. Let everyone else be equally *wowzer*-ed by Hydra-Zen. :dragon: :dragon_face:

**Reply by rsokl** [reply]
So sorry for the late-late-late reply! Happy new year 🥳 ! I am very excited to see hydra-zen's branding on your **now over 1K ⭐** repo!! (congrats!!)

Regarding building a bridge between Hydra and Click, I will have to look into that. It could be complicated...LY COOL! But, for now, I am still working on building up the core functionality & docs of hydra-zen, and then hope to post some articles...somewhere? Medium? Towards Data Science? to get an actual userbase. (I am pretty confident that many Hydra users would ultimately prefer using hydra-zen rather than hand-crampingly write repetitive configs themselves, but I just haven't found a way to effectively get the word out there).

Perhaps once I do amass a more substantial user base, I will have some eager contributors who can look into devouring a bunch of other APIs.

Sorry again for falling off the radar. I hope all is well!

**Comment by leycec** [comment]
:nerd_face: 

> I am very excited to see hydra-zen's branding on your now over 1K star repo!! (congrats!!)

**Aww, yeah!** You *sooo* nice for noticing our little social milestone.

I also now feel like a proper jerk, because the current icon showcase (including your *wonderfully* mythic icon) might be going away shortly. And... this suddenly became an awkward moment, didn't it? `</gulp>`

Why is this bad thing happening? Because I'll *probably* be throwing up a shamelessly monetized set of GitHub Sponsor tiers in the wake of our upcoming `beartype 0.10.0` tiger drop – and <sup>wait for it</sup> the highest tier will *probably* be modestly paid advertisement of a corporate or project icon on our frontpage. <sup>*Nooooooooooooo!*</sup>

I'm *not* a capitalist kinda guy. I'm a cabin-in-the-woods kinda guy. Ain't nobody in the woods for the :dollar: on delivery. So, this is as awkward for me as it is for everyone else.

Unfortunately, my beautiful wife has kindly – but firmly and *very* repeatedly – informed me that @beartype gotta metastasize. Us previously happy-go-lucky and plucky Canadians have been steamrolled over by livid increases across the consumer price index. @beartype now needs to pull its hefty two-ton weight by paying for monthly cat food. *Curse those fragile feline bodies.*

I don't like this – because it really feels like a betrayal of my open-source volunteer values. Nobody gets into GitHub for that creepy eye in the pyramid. But I like feeding our cats. The tension is palpable.

I'm *so* sorry. :disappointed: 

> ...but I just haven't found a way to effectively get the word out there.

**Word, yo.** As a died-in-the-wool neckbeard who only wears black wool (*...less fun in August than it seems*), I'm innately adverse to social media and self-promotion. I only vaguely know what Twitter, Instagram, and TikTok are. They're just something ever *there.*

Because I voluntarily eat glutin-free muesli, @beartype was once nothing. Then I randomly posted our GitHub frontpage on [a lazy-eyed Sunday](https://www.delayforreddit.com/analysis/subreddit/python) to both [/r/Python](https://reddit.com/r/python) and [Hacker News](https://news.ycombinator.com). Then @beartype farmed 300 GitHub stars in 3 minutes.

That's when I realized that *our* kind of social media might not be so bad after all. I have no useful advice – only ancient wisdom gleaned from my extroverted betters:

1. Maybe author articles as (sub)sections of your existing `readthedocs.org`-hosted documentation? Own your brilliance.
1. Submit permalinks to those articles to both [/r/Python](https://reddit.com/r/python) and [Hacker News](https://news.ycombinator.com).
1. Farm 300 GitHub stars in 3 minutes.
1. Chuckle mirthlessly as you tweet your first selfie before your voluminous GitHub stargazer list that's expanding as fast as the known universe into speculative and frankly suspicious dark matter.

> I hope all is well!

Oh, you *so* gracious! My beautiful wife, my blotchy neckbeard, and our two mewling cats all hope that 2022 also finds you hip-deep in exploding stars, unending hype trains, and a positive conflagration of social media and bling-bling money shots. :star2:  :star_struck: :star2:

**Reply by posita** [reply]
> I'm _not_ a capitalist kinda guy. I'm a cabin-in-the-woods kinda guy. Ain't nobody in the woods for the 💵 on delivery. So, this is as awkward for me as it is for everyone else.
> 
> Unfortunately, my beautiful wife has kindly – but firmly and _very_ repeatedly – informed me that @beartype gotta metastasize. Us previously happy-go-lucky and plucky Canadians have been steamrolled over by livid increases across the consumer price index. @beartype now needs to pull its hefty two-ton weight by paying for monthly cat food. _Curse those fragile feline bodies._
> 
> I don't like this – because it really feels like a betrayal of my open-source volunteer values. Nobody gets into GitHub for that creepy eye in the pyramid. But I like feeding our cats. The tension is palpable.

You kin tell Mrs. @leycec that I tol' you tuh quit whinin' and put out a tip jar, or a GoKickMe page, or tiered advertisin' thingamajig, or whatever. Ain't nuthin' soul-corruptin' 'bout it. Cats gotta eat, so does she, and so do you. That neck beard, wool, an' pellet stove ain't gonna keep your core temperature up by themselves, nor your lady's neither. (I'm assuming the wife don't have the insulatin' benefits of a neck beard, but no judgment either way.) If folks like what yer doin', they should be able to pay ya for it, plain and simple. Ya can't go 'round denyin' 'em that. Heck, it may even help ya keep doin' it. Ever'body wins. That's jes mah two cents, which is worth what ya paid fer it.

![Hat tip](https://c.tenor.com/9x7sGKcXKBAAAAAM/the-big-lebowski-the-stranger.gif)

**Comment by rtbs-dev** [comment]
Not to, yknow, snoop on (or necro) a fascinating conversation or anything, but it's amazing to me how disparate projects I've followed end up finding each other! :sweat_smile: 

I noticed: 
> The argparse.ArgumentParser API makes me want to rub chalk into my eyes, whereas [Click](https://click.palletsprojects.com/) rocks the POSIX Casbah but fails to integrate the CLI APIs it auto-generates with configuration files – which is always the pain point. Just "Ugh!"

So... have we all tried [`Typer`](https://typer.tiangolo.com/)? It's all built on click iirc, and uses _gasp_ type-hints to auto-produce documentation and flag vs. arg parsing, etc.! I was literally just looking at my shiny new toys between Hydra, beartype, and Typer CLI, and thinking to myself... "great, _now_ how do we all play in the sandbox together nicely?" And also "how do I do dependency injection for my typer application now?" Low-and-behold, hydra-zen _proceeds to exist_. 

So, yeah, that's something I'm working on

**Reply by rsokl** [reply]
> So... have we all tried [Typer](https://typer.tiangolo.com/)?

I haven't!

> looking at my shiny new toys between Hydra, beartype, and Typer CLI, and thinking to myself... "great, now how do we all play in the sandbox together nicely?" And also "how do I do dependency injection for my typer application now?" Low-and-behold, hydra-zen proceeds to exist.... So, yeah, that's something I'm working on

*awesome* 🤩 

Wow, what timing. I had *just today* been thinking back to this thread, and how nice it would be to follow @leycec 's advice and gobble up click to give Hydra a nicer CLI. I *also* had just been reflecting on the potential of extending hydra-zen's auto-config capabilities based on fields' type-annotations... and then I see your post!

I think that there as some real low-hanging fruit to be had here. I'd love to stay posted on your progress and would be happy two share my two cents where they might be worth something. If there is anything on the hydra-zen side -- explanations or potential new features -- please let me know. Feel free to swing by [our discussion board](https://github.com/mit-ll-responsible-ai/hydra-zen/discussions)!

**Comment by leycec** [comment]
**Fascinating discussion continues fascinating.** You're blowing my puny mind, everybody. On the cusp of well-typed CLI API greatness, let us all take a moment and reflect on @beartype's growing fascination with [statement-level runtime type-checkers like `beartype.abby.is_bearable()` and `die_if_unbearable()`](https://github.com/beartype/beartype#beartype-at-any-time-api).

I dimly recall @rsokl pining for something like those functions... like, *really* dimly. So maybe it didn't happen. But I'm pretty sure it happened.

Previously, `is_bearable()` exhibited embarrassingly bad runtime performance in common edge cases. Now and in our upcoming `beartype 0.11.0` release, `is_bearable()` enjoys the same `O(1)` "constant time with negligible constants" guarantee as `@beartype`-decorated callables. Let's type-check a **list of one million 2-tuples of NumPy arrays** in nearly **1µs** (i.e., one millionth of a second):

```python
>>> from beartype.abby import is_bearable
>>> from beartype.typing import List, Tuple
>>> import numpy as np
>>> data = [(np.array(i), np.array(i+1)) for i in range(1000000)]
>>> %time is_bearable(data, List[Tuple[np.array, np.array]])
       CPU times: user 31 µs, sys: 2 µs, total: 33 µs
       Wall time: 36.7 µs
False
```

Let the well-typed CLI APIs flow like fine Canadian apple cider, I say. :clinking_glasses:

**Reply by rsokl** [reply]
>let us all take a moment and reflect on https://github.com/beartype's growing fascination with [statement-level runtime type-checkers like beartype.abby.is_bearable() and die_if_unbearable()](https://github.com/beartype/beartype#beartype-at-any-time-api).

OOOOOooooh! Awesome! This is a really nice piece of functionality. Thank you so much for letting me know about these functions! 🙏 🙏 🙏
